### ISSUE TITLE
`git town set-parent` suddenly asks for parent information for other branches

### ISSUE THREAD

**Issue by ruudk** [git-town/git-town#3676]
### Steps to reproduce

Today I upgraded from 14.2.0 to 14.2.2.

I noticed an odd behavior today.

When I ran `git town set-parent` it prompted me for a parent. I run this command a lot so I did it without taking care what it asked. It then prompted me again. 

When I read closely, I noticed that it asked me to setup parent branches for other branches that I had recently checked out. Branches from co-workers. But it also asked me for parent info for very old stale branches.

It seems that something is wrong with this command.

### Command output

```shell
I don't want to post the `--verbose` output here as it contains a lot of sensitive information (branch names from the past years).


❯ git rev-parse --abbrev-ref HEAD
ruudk/2024/6/remove-TestMessage

❯ git town set-parent

Parent branch for experiment/bash-assert

Please select the parent of branch "experiment/bash-assert" or enter its number.
Most of the time this is the main development branch (main).


00   <none> (perennial branch)
01 > main
02 > .. more branches
```
```


### Additional information

_No response_

**Comment by kevgo** [comment]
Thanks for submitting this. Did you run this in a subfolder of your Git repository by any chance?

**Comment by ruudk** [comment]
I checked my terminal history and I was not in the wrong branch.

**Comment by ruudk** [comment]
This morning, the same thing happened, when I resolved a merge conflict while running `sync`. I then used `continue` and got this:

```
❯ sync

[ruudk/2024/6/rector-on-ci] git fetch --prune --tags

[ruudk/2024/6/rector-on-ci] git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

[main] git rebase origin/main
Current branch main is up to date.

[main] git checkout ruudk/2024/6/rector-on-ci
Switched to branch 'ruudk/2024/6/rector-on-ci'
Your branch is up to date with 'origin/ruudk/2024/6/rector-on-ci'.

[ruudk/2024/6/rector-on-ci] git rebase main
Auto-merging phpstan-baseline.php
CONFLICT (content): Merge conflict in phpstan-baseline.php
error: could not apply 376b15db8af... Apply Rector fixes
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 376b15db8af... Apply Rector fixes


Error: exit status 1

To continue after having resolved conflicts, run "git town continue".
To go back to where you started, run "git town undo".
To continue by skipping the current branch, run "git town skip".

❯ git town continue

Parent branch for experiment/bash-assert

Please select the parent of branch "experiment/bash-assert" or enter its number.
Most of the time this is the main development branch (main).


00   <none> (perennial branch)
01 > main
```

**Comment by kevgo** [comment]
Thanks for the additional details. Can you please share how many branches you have in total, i.e. what is the output of this command on your machine?

```bash
git branch | wc -l
```

> When I ran `git town set-parent` it prompted me for a parent. I run this command a lot 

Can you please share why you run `git town set-parent` so much? This command exists only to fix broken configuration manually. What is your workflow for working with branches?

> it asked me to setup parent branches for other branches that I had recently checked out. Branches from co-workers

Are you using the commands `git town observe` and `git town contribute` to check out these branches or mark them as observed/contribute branches?

> it also asked me for parent info for very old stale branches.

Same here, are you using `git town park` to tell Git Town that these branches are intentionally stale?

**Comment by kevgo** [comment]
Also, what is the naming structure of your branches? I already know that they contain slashes. Do their names also contain periods or other special characters?

**Comment by kevgo** [comment]
I have just released [v14.2.3](https://github.com/git-town/git-town/releases/tag/v14.2.3) that contains a fix for my currently only hypothesis for this cause of this problem.

**Comment by ruudk** [comment]
```command
$ git branch | wc -l
      96
```

I used `git town set-parent` to manually set another branch as parent. Sometimes I use `hack` to create a branch, and then later realize I want to make this a parent for another branch. 

> Are you using the commands git town observe and git town contribute to check out these branches or mark them as observed/contribute branches?
I'm not. Sometimes I checkout PR's of other team members (or Dependabot) and simply use `gh` for this:
![Screenshot 2024-06-26 at 08 32 26@2x](https://github.com/git-town/git-town/assets/104180/3fc2cad6-2b44-4e4f-bd2d-d73f132f165b)

So I never used the observe/contribute features.

I'm also not using `git town park`. Maybe I should use that for now to mitigate this issue.

Could it be that for a long time Town did not care about branches without a parent? And that since a few versions ago, it did. And therefore wants me to assign parents for all those old branches that it previously did not bother with?


> Also, what is the naming structure of your branches? I already know that they contain slashes. Do their names also contain periods or other special characters?

It's mostly `ruudk/2024/6/some-FeatureName123`.

> I have just released [v14.2.3](https://github.com/git-town/git-town/releases/tag/v14.2.3) that contains a fix for my currently only hypothesis for this cause of this problem.

I tried but it did not do much:

```
❯ git rev-parse --abbrev-ref HEAD
ruudk/2024/6/rector-small-things

❯ git town set-parent

Parent branch for ruudk/2022/12/decorate-translation-twig-function

Please select the parent of branch "ruudk/2022/12/decorate-translation-twig-function" or enter its number.
Most of the time this is the main branch (main).


00   <none> (perennial branch)
01 > main
```

As you can see, it wants me to assign a parent for a VERY old branch that I never synced. But still want to keep.

**Comment by ruudk** [comment]
I just went through all my old branches, using `git town set-parent` and set the main branch as parent. I think this will solve it. No idea how this suddenly became an issue.

**Comment by kevgo** [comment]
> 96 local branches

I'll add some end-to-end test to make sure Git Town and the underlying Git configuration database scale to so many branches.

> Sometimes I checkout PR's of other team members (or Dependabot) and simply use `gh` for this:

That's a good use case that I haven't thought about much. Yes, if you regularly checkout other people's branches using external tools like `gh` then Git Town will keep asking for their parent. The big missing feature here is #2745: Git Town could use the parent branch of the pull request as the local parent, and therefore doesn't need to ask you.

> Could it be that for a long time Town did not care about branches without a parent? And that since a few versions ago, it did. And therefore wants me to assign parents for all those old branches that it previously did not bother with?

Yes. In the last couple of releases I "tightened up" Git Town's data management to be more accurate and strict. As a result, a lot of problems (like unknown parent branches) that were silently ignored previously now get surfaced.

**Comment by ruudk** [comment]
So after I set all the parents for those old branches, I did not encounter any issues today. That means that the issue is resolved and it was due the fact that for some reason those old branches did not have a parent.

Thanks for your detailed response and keep up the amazing work 💙 Is there a sponsor page somewhere?

**Comment by kevgo** [comment]
I'm glad we got this one resolved! I have just added a sponsor link for Git Town. It's on the Git Town repository page on GitHub or [here](https://github.com/sponsors/git-town).

**Comment by ruudk** [comment]
@kevgo Found it!

**Comment by kevgo** [comment]
Wow, thanks!! 🙏

**Comment by kevgo** [comment]
This is now available in [v16.3](https://github.com/git-town/git-town/releases/tag/v16.3.0).
